mozilla/kitsune
# [Bug 1166090] Document template properties validation

This makes some important, but hopefully pretty straight forward changes to the Document model, and then changes a whole lot of tests to account for the model changes.

* Added FactoryBoy powered model factories. These have some niceties over model makers, and Fjord uses them too.
* Enforce that documents with a title like "Template:Foo" are in the Template category, and vice-versa.
* Update all the tests to be ok with that, generally by switching them from model makers to model Factories.
* Make renaming a template in such a way that it is no longer a template not create redirects. The automatic redirect creation logic doesn't work when templates are required to have "Template:" prefixes *and* be in the right category.

To test:

* Try and rename a template to not have the prefix (without changing the category). It should fail with a nice error message. (There is a test for this)
* Try and change the category of a template to something else (without changing the title). It should fail. (There is a test for this)
* Try and do both at once. It should work.  (There is a test for this)
* Try and do this for a localized article. This is where most of our real problems are. (I need to add a test for this).

TODO

I realized I have a few more things to do before I want to land this. But it should be ready to start review now.

* [x] Add test for localized articles.
* [x] Add tests for normal article -> templates
* [x] ~~Figure out if landing this on a DB full of title/category inconsistencies is going to be a bad idea. I suspect it will only be a problem if the document is edited, in which case it will give a nice error to a human. But maybe not? What else calls .save() on documents? Hunt these down.~~
* [x] Wait for the templates on the site to be renamed so landing this won't break everything.

(This isn't ready to merge yet, because of the above)

f?
